### PR TITLE
Remove SciPy from the install requirement

### LINKED ISSUE #3035: Lightweight install option?
I'd like to use Optuna for a project, but the installed package with dependencies comes to a whopping 174MB. This is an issue for my application since I'd like to use it in AWS Lambda, which has a hard cap of 250MB for all code in a lambda, and some other packages I'm using consume 90MB (bringing me to 260MB).

Is there a lightweight install option for Optuna, a way to strip out any optional functionality? Or any other suggestions on how to get the size down?

Thanks!

### PR DESCRIPTION
**Pull request by not522**
## Motivation
SciPy requires relatively large storage and memory. It causes trouble in some constrained environments. E.g., This PR resolves #3035.

The refactoring of `setup.py` is going in #3517. I will update this PR after #3517 is merged.

## Description of the changes
Remove SciPy from the install requirement

### CODE REVIEW

**Comment by c-bata** [comment]
@HideakiImamura @contramundum53 Could you review this PR?

**Comment by not522** [comment]
My implementation of truncnorm is too slow in some cases, and it causes the CI to get stuck. I'm working to fix it.

File: optuna/samplers/_tpe/parzen_estimator.py [hunk 1]
```diff
@@ -448,7 +436,7 @@ def _normal_cdf(x: np.ndarray, mu: np.ndarray, sigma: np.ndarray) -> np.ndarray:
         denominator = x - mu
         numerator = np.maximum(np.sqrt(2) * sigma, EPS)
         z = denominator / numerator
-        return 0.5 * (1 + special.erf(z))
+        return 0.5 * (1 + np.vectorize(math.erf)(z))
```
Comment (contramundum53) [inline optuna/samplers/_tpe/parzen_estimator.py:439]: FYI, I micro-benchmarked this line and got the following results:
```
In [10]: xs = np.linspace(-2, 2, 100000)

In [11]: %timeit np.vectorize(math.erf)(xs)
6.5 ms ± 35.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [12]: %timeit special.erf(xs)
631 µs ± 36.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

`np.vectorize(math.erf)` could be 10 times slower than `special.erf`. This could be or not be a problem.

File: optuna/samplers/_tpe/parzen_estimator.py [hunk 2]
```diff
@@ -200,8 +186,10 @@ def log_pdf(self, samples_dict: Dict[str, np.ndarray]) -> np.ndarray:
                     )
                     log_pdf = np.log(cdf + EPS) - np.log(p_accept + EPS)
             component_log_pdf += log_pdf
-        ret = special.logsumexp(component_log_pdf + np.log(self._weights), axis=1)
-        return ret
+        weighted_log_pdf = component_log_pdf + np.log(self._weights)
+        max_ = weighted_log_pdf.max(axis=1)
+        with np.errstate(divide="ignore"):
```
Comment (contramundum53) [inline optuna/samplers/_tpe/parzen_estimator.py:191]: Why was this necessary, when there doesn't seem to be a case where division by zero is involved?

Comment (not522) [reply optuna/samplers/_tpe/parzen_estimator.py:191]: It follows the SciPy's implementation. I will add a comment why `np.errstate` is necessary. https://github.com/scipy/scipy/blob/7076a90ff42a2349c621e1a081fbf5f715c421c0/scipy/special/_logsumexp.py#L112-L113

Comment (contramundum53) [reply optuna/samplers/_tpe/parzen_estimator.py:191]: Is there a case where that would incur a division by zero?
(Except when `self._weights` are all zero, which would be invalid anyways)

Comment (not522) [reply optuna/samplers/_tpe/parzen_estimator.py:191]: It introduced to resolve inf related bug. It never occurs in our situation, so we can remove `np.errstate`. Thank you for careful review!
https://github.com/scipy/scipy/issues/4014
https://github.com/scipy/scipy/pull/4016

Comment (not522) [reply optuna/samplers/_tpe/parzen_estimator.py:191]: By the way, `np.errstate(under='ignore')` may be preferred. #2639

**Comment by contramundum53** [comment]
Other than my comment and https://github.com/optuna/optuna/pull/3560#issuecomment-1127231048, this PR looks good to me. I will review this again after https://github.com/optuna/optuna/pull/3560#issuecomment-1127231048 is fixed.

**Comment by contramundum53** [comment]
For https://github.com/optuna/optuna/pull/3560#issuecomment-1127231048, I think sampling from a uniform distribution and evaluating its inverse error function will suffice. However, I could not find an implementation of inverse error function without relying on `scipy.special.erfinv`, so I think we need to bring its implementation to optuna (which can be a little tough).

**Comment by contramundum53** [comment]
I took a benchmark of the above code, compared with scipy.
```
l, r = -0.001, 0.001
sample_trunc_normal(l, r, size=1000, oversample_rate=2) 
# ==> 62.8 µs ± 279 ns
stats.truncnorm(l, r).rvs(size=1000)
# ==> 316 µs ± 19 µs
```
